### PR TITLE
Refactor PermissionLogController tests

### LINKED ISSUE #1827: Refactor PermissionLogController tests
Once we've copied the code and tests for PermissionLogController from the extension, and after redesigning the API, we will want to rewrite the tests so that they conform to our [unit testing guidelines](https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md). Primarily this means that we will want to organize tests by method.

### PR DESCRIPTION
**Pull request by cryptodev-2s**
## Explanation

This pull request is designed to align the tests for `PermissionLogController` with the guidelines outlined in our contributor documentation. The objective is to refactor these tests in accordance with best practices recommended for unit testing, as detailed in the provided documentation [link](https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md).

## References

* Fixes #1827

## Changes
- Discontinued the use of beforeEach for the majority of the tests, with exceptions made for certain tests that utilize fake timers. Now, each test individually initializes a new controller and middleware within its own scope, enhancing test isolation and clarity.
- Split long test scenarios for more focused test cases.
## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/permission-log-controller/tests/PermissionLogController.test.ts [hunk 1]
```diff
@@ -366,97 +368,101 @@ describe('PermissionLogController', () => {
...
         expect(controller.state.permissionActivityLog).toHaveLength(0);
       });
 
-      it('enforces log limit', () => {
-        const req = RPC_REQUESTS.test_method(SUBJECTS.a.origin);
-        const res = {
-          ...PendingJsonRpcResponseStruct.TYPE,
-          result: ['bar'],
-        };
+      describe('log limit enforcement', () => {
```
Comment (mcmire) [inline packages/permission-log-controller/tests/PermissionLogController.test.ts:394]: In the [unit testing contributor guide](https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md#-use-describe-to-group-tests-under-scenarios) I recommended that we use `describe` not to group together tests by behavior (what the thing we're testing does), but rather by scenario (_why_ the thing we're testing is doing what it's doing). I know different people have different ways of structuring tests, so there's not necessarily one right answer here. However, I feel that grouping tests in this way makes it easier to compare tests to implementation to ensure that you have complete coverage. The only real way to know whether you have complete coverage for a piece of code is to map out all of the code paths (all of the routes that the code will take through different branches of logic). When mapping out the code paths you end up producing a list that looks something like this:

- When X is true and Y is true: do something
- When X is true and Y is false: do something else
- When X is false: do yet another thing

So, structuring tests by using `describe` to describe the context that creates the behavior (the part before the ":" here) and then using `it` to describe the behavior (the part afterward) is a natural outcome of this exercise:

``` typescript
describe('when X is true and Y is true', () => {
  it('does something', () => {
    // ...
  });
});

describe('when X is true and Y is false', () => {
  it('does something else', () => {
    // ...
  });
});

describe('when X is false', () => {
  it('does yet another thing', () => {
    // ...
  });
});
```

Considering the way that these tests are presently written, using this structure throughout this file would result in rewriting it completely. However, perhaps it's worth it to nudge them in that direction? In other words, maybe you don't need this one `describe` in particular?

What are your thoughts?

File: packages/permission-log-controller/tests/PermissionLogController.test.ts [hunk 2]
```diff
@@ -366,97 +368,101 @@ describe('PermissionLogController', () => {
...
-          LOG_METHOD_TYPES.restricted,
-          true,
-        );
+        it('removes the oldest log entry when a new one is added after reaching the limit', () => {
+          const controller = initController();
+          const logMiddleware = initMiddleware(controller);
+          const req = RPC_REQUESTS.test_method(SUBJECTS.a.origin);
+          const res = { ...PendingJsonRpcResponseStruct.TYPE, result: ['bar'] };
+
+          // Initially fill the log to its limit
```
Comment (mcmire) [inline packages/permission-log-controller/tests/PermissionLogController.test.ts:416]: Nit: I feel like if we make the code clear enough, the variable names readable enough, the test is short enough, it's easy enough to follow along without needing comments to describe each test. What are your thoughts on removing them?

**Review by mcmire — COMMENTED** [review]
It seems that this controller is more complex than I had anticipated. If you look at `#logRequest`, `#getRequestedMethods`, `#logResponse`, and `#logPermissionsHistory` you will see that there is an incredible amount of logic hidden in the middleware. I don't think we are testing all of the code paths; however, in order to achieve true coverage, I think we would have to rewrite this entire file, which seems out of scope for this ticket. Given this I tried to leave some comments that seem easier to tackle. I think the other changes you've made here make sense, though, and we can spend more time later making this better.

File: packages/permission-log-controller/tests/PermissionLogController.test.ts [hunk 3]
```diff
@@ -112,134 +111,135 @@ function validateActivityEntry(
...
-        res = {
-          ...PendingJsonRpcResponseStruct.TYPE,
-          result: ACCOUNTS.c.permitted,
-        };
-
-        logMiddleware(req, res, mockNext, noop);
+      describe('recording Activity for restricted methods', () => {
+        it('records activity for a successful restricted method request', () => {
+          const controller = initController();
+          const logMiddleware = initMiddleware(controller);
```
[the remark this answers is not in the document]
Comment (mcmire) [reply packages/permission-log-controller/tests/PermissionLogController.test.ts:135]: I'm curious whether we even need this function, if we can just call `controller.createMiddleware()` directly without having to wrap it.

File: packages/permission-log-controller/tests/PermissionLogController.test.ts [hunk 4]
```diff
@@ -112,134 +111,135 @@ function validateActivityEntry(
...
+          const req = RPC_REQUESTS.test_method(SUBJECTS.a.origin);
```
[the remark this answers is not in the document]
Comment (cryptodev-2s) [reply packages/permission-log-controller/tests/PermissionLogController.test.ts]: @mcmire good point for clarification at least we should leverage specifying restrictedMethods upon each controller initialisation to have a broader view of how it should work.
I just updated it to specify both state and restrictedMethods

**Review by mcmire — APPROVED** [review]
Looks good!

**Pull request merged**